allow encoding literal JSON values

aSemy opened the pull request:
This PR provides a new function for encoding raw JSON content, without quoting it as a string. This allows for encoding JSON numbers of any size or precision, so BigDecimal and BigInteger can be supported.

[Link to view the updated documentation](https://github.com/aSemy/kotlinx.serialization/blob/feat/1051-big_decimals/docs/json.md#encoding-raw-literal-json-content-experimental)

fix: #1051 
fix: #1405

related: 

* #1557
* https://github.com/ionspin/kotlin-multiplatform-bignum/issues/219

The implementation is similar to how unsigned numbers are handled.

* `JsonUnquotedLiteral()` is a new function that allows creating literal JSON content.
* I added `val coerceToInlineType` to `JsonLiteral`, so that `JsonUnquotedLiteral` could use `encodeInline()` 
* I defined `val jsonUnquotedLiteralDescriptor` as a 'marker', for use with `encodeInline()`
* `ComposerForUnquotedLiterals` (based on `ComposerForUnsignedNumbers`) will 'override' the encoder when a `JsonLiteral` has the `jsonUnquotedLiteralDescriptor` marker, and will encode the content as a string *without* surrounding quotes.

~I'm not sure how to proceed with decoding literal numbers.~ I just realised, there's no need for special implementation - primitives are all decoded the same.

aSemy commented:
I think more tests and specialist knowledge is needed, but the approach is ready for review. I'm really keen to provide a way of supporting BigDecimal, as it's been a blocker in a couple of my projects for a long time.

YanishR reviewed (commented):
When do you guys foresee there will be progress on this? I have a use case for it in my project

sandwwraith commented:
Hi, thanks for your PR! I think I have time for it on this week, stay tuned

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -61,6 +63,17 @@ public fun JsonPrimitive(value: Number?): JsonPrimitive {
+ */
+public fun JsonLiteralNumber(value: String?): JsonPrimitive {
+    if (value == null) return JsonNull
+    if (!jsonLiteralNumberRegex.matches(value)) throw JsonEncodingException("not a number: $value")
```
Regex validating would block various other use-cases, like the one mentioned here: https://github.com/Kotlin/kotlinx.serialization/issues/1405#issuecomment-813381394

Now I'm a bit hesitant about whether we should support this (because we can't parse it back as the same raw value, but it is not needed), but it looks like a good use case. Supporting it would also require renaming this function into something like `JsonUnquotedLiteral` to avoid any connection to numbers, so it's better to think about this from the beginning.

aSemy replied:
In order to keep this PR focused I just wanted to completely prevent encoding anything except JSON numbers. I thought that allowing encoding raw strings would be a much larger discussion that would prevent an incremental improvement. Of course, changing this PR to allow for encoding raw strings is also possible!

sandwwraith commented:
I think we should have additional design discussion on a general approach (using `encodeInline` and `JsonPrimitive(isString = true` instead of a separate `JsonEncoder.encodeUnquoted` function); I'll keep you posted

aSemy commented on formats/json/commonMain/src/kotlinx/serialization/json/internal/Composers.kt:
```diff
@@ -60,6 +60,13 @@ internal class ComposerForUnsignedNumbers(writer: JsonWriter, private val forceQ
+@SuppressAnimalSniffer
+internal class ComposerForLiteralNumbers(writer: JsonWriter, private val forceQuoting: Boolean) : Composer(writer) {
+    override fun printQuoted(value: String) {
+        if (forceQuoting) printQuoted(value) else super.print(value)
```
thanks, fixed by changing `printQuoted(value)` to `super.printQuoted(value)`

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/internal/Composers.kt:
```diff
@@ -60,6 +60,13 @@ internal class ComposerForUnsignedNumbers(writer: JsonWriter, private val forceQ
     }
 }
 
+@SuppressAnimalSniffer
```
AnimalSniffer is a plugin that detects and reports usages of Java 8 API. It is useful for checking if your library can run on old Android. Now, given that the default Kotlin target is 1.8 and Android has desugaring, everything still works, but the plugin still complains.

You should move annotation to `TreeJsonEncoder.inlineUnsignedNumberEncoder` from `TreeJsonEncoder.encodeTaggedInline`

aSemy commented on formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonEncoder.kt:
```diff
@@ -169,6 +173,14 @@ internal class StreamingJsonEncoder(
         else ComposerForUnsignedNumbers(composer.writer, forceQuoting)
     }
 
+    private fun composerForLiteralNumbers(): Composer {
```
I think that works well 👍 . I've implemented it.

aSemy commented on formats/json-tests/jvmTest/src/kotlinx/serialization/BigDecimalTest.kt:
```diff
@@ -0,0 +1,220 @@
+}
+
+@Serializable
+private data class BigDecimalHolder(
+    @Contextual
+    val bd: BigDecimal,
+    @Contextual
+    val bdList: List<@Contextual BigDecimal>,
+    val bdMap: Map<@Contextual BigDecimal, @Contextual BigDecimal>,
+)
```
I think the repetition of `@Contextual` here would be reduced by https://github.com/Kotlin/kotlinx.serialization/issues/507. (Of course here this can be set with a file-annotation, but in other codebases it's likely that BigDecimal is used in many, many files.

sandwwraith reviewed (changes requested):
After a discussion, the following points were established:

* Function should accept arbitrary strings and do not validate them with regex. This will allow users to do various json-embeddings, support non-base10 numbers, different notations, etc and will also simplify an invariant for users and save us from potential problems with regex correctness and performance.

* Function should accept a `String?` and return `JsonNull` if a null was passed to it, for consistency with `JsonPrimitive` overloads.
* However, if string literal `"null"` is passed (i.e. `value == "null"`), function should throw `IllegalArgumentException`. This is motivated by the fact that when we serialize such a literal, if effective becomes indistinguishable from `JsonNull` and can't be deserialized back to `JsonLiteral`. We cannot know for sure if this was a user intent to do so or bug in validation. Better to prohibit such cases. IAE message can sound like this: "It is impossible to create a literal unquoted value of 'null'. If you want to create json null literal, use JsonNull object, otherwise, use quoted literal." 

Summarising this all up, I suggest to name the new function 'JsonUnquotedPrimitive'. What do you think?

aSemy commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -105,6 +126,14 @@ internal class JsonLiteral internal constructor(
 }
 
+/** Used as a marker to indicate the encoder should use `encodeInline()` */
+internal val jsonLiteralNumberDescriptor: SerialDescriptor = JsonLiteralNumber.serializer().descriptor
```
1. ✅
2. `InlinePrimitiveDescriptor` is internal to kxs.core, so it can't be accessed in kxs.json. https://github.com/Kotlin/kotlinx.serialization/blob/daa95c79ffadc0eedbbb4a481a00556b78212e43/core/commonMain/src/kotlinx/serialization/internal/InlineClassDescriptor.kt#L28

sandwwraith replied:
It worth changing to `@InternalSerializationApi public fun InlinePrimitiveDescriptor` then

aSemy commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -61,6 +63,17 @@ public fun JsonPrimitive(value: Number?): JsonPrimitive {
+/**
+ * Creates [JsonPrimitive] from the given number.
+ */
+public fun JsonLiteralNumber(value: String?): JsonPrimitive {
```
@sandwwraith Regarding the rename, you suggested `JsonUnquotedPrimitive`. However, the provided value might not be a primitive (users could use it to encode JSON arrays or objects), and the because might have quotes it could be a bit misleading - the new function doesn't really care about quotes.

How about `JsonLiteralElement` or `JsonRawElement`?

sandwwraith replied:
Well, the 'unquoted' part simply means that we won't put quotes around the value. If one creates `JsonUnquotedPrimitive("\"foo\"")` of course it would be quoted — but that is defined by the data, not the function behavior.

> However, the provided value might not be a primitive 

Yes, that's a valid point. Maybe `JsonUnquotedLiteral` then?

aSemy replied:
I'd prefer having 'Raw' because it matches similar APIs in other serialization libraries, and it's a common word ('raw' is mentioned in each of the tickets I linked in the PR description), but that's minor - `JsonUnquotedLiteral` is fine.

I have another thought though. What about providing a function overload for `JsonPrimitive()`?

```kotlin
/**
 * Creates [JsonPrimitive] from the given string.
 *
 *  @param[quoted] If false, the value will not be wrapped in quotes <insert more docs here>
 */
@ExperimentalSerializationApi
public fun JsonPrimitive(value: String?, quoted: Boolean): JsonPrimitive {
    return if (!quoted) JsonPrimitive(value)
    else when (value) {
        null -> JsonNull
        JsonNull.content -> throw JsonEncodingException("It is impossible to create a literal unquoted value of 'null'. If you want to create JSON null literal, use JsonNull object, otherwise, use JsonPrimitive")
        else -> JsonLiteral(value, isString = false, coerceToInlineType = jsonRawElementDescriptor)
    }
}
```

It's a little odd, because `@ExperimentalSerializationApi` means it has to be a separate function rather than adding a new argument with a default to the existing `JsonPrimitive()`, but if this gets promoted to stable, then the function could be merged.


```kotlin
public fun JsonPrimitive(value: String?, quoted: Boolean = true): JsonPrimitive {
  // as above
}
```

aSemy commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -61,6 +63,20 @@ public fun JsonPrimitive(value: Number?): JsonPrimitive {
+ * // TODO add more docs once functionality is stabilised.
+ */
+@ExperimentalSerializationApi
+public fun JsonLiteralNumber(value: String?): JsonPrimitive {
```
> However, if string literal "null" is passed (i.e. value == "null"), function should throw IllegalArgumentException. This is motivated by the fact that when we serialize such a literal, if effective becomes indistinguishable from JsonNull and can't be deserialized back to JsonLiteral. We cannot know for sure if this was a user intent to do so or bug in validation. Better to prohibit such cases. IAE message can sound like this: "It is impossible to create a literal unquoted value of 'null'. If you want to create json null literal, use JsonNull object, otherwise, use quoted literal."

@sandwwraith Could you explain more about the above, please? I don't understand how a problem could occur.

I thought maybe it was because serialisation should be symmetric, so all `JsonPrimitive`s should decode to `JsonPrimitive`. But `JsonNull` is a subtype of `JsonPrimitive`, so that seems like it works to me.

qwwdfsad replied:
Hi!

Our concern is not about `null` and resulting `JsonNull`, but `"null"` *string* literal. 

The idea is the following -- with this PR, we enable bidirectional _unquoted_ string literal, meaning that we can both write them to `JsonElement` and parse them back.

The problem is simple -- in unquoted mode, both `null` and `"null"` have *the same* serialized form (`{"key": null}`) and cannot be distinguished when read back. It means that we have the following potential options:

* Allow both literals for writing (creating programmatically), but both of them as `JsonNull`. This may lead to an inconsistent data model, but it is consistent with `lenient` mode (https://github.com/Kotlin/kotlinx.serialization/issues/1600)
* Prohibit `"null"` literals as indistinguishable when reading it back. The possible workaround is to use quoted mode for such literals.

We'd like to stick with the second option, but we are open for discussion

aSemy replied:
Sorry, I don't really follow. I'm asking so I can figure out how to write a unit test for this, because at the moment if I disable the exception

```kotlin
@ExperimentalSerializationApi
@Suppress("FunctionName")
public fun JsonUnquotedLiteral(value: String?): JsonPrimitive {
    return when (value) {
        null -> JsonNull
//        JsonNull.content -> throw JsonEncodingException("It is impossible to create a literal unquoted value of 'null'. If you want to create JSON null literal, use JsonNull object, otherwise, use JsonPrimitive")
        else -> JsonLiteral(value, isString = false, coerceToInlineType = jsonUnquotedLiteralDescriptor)
    }
}
```

none of the test fail, e.g. https://github.com/Kotlin/kotlinx.serialization/commit/d00bff2bbda69a17083480ec9daaaad5a54d2b42 (except for the test that is checking that `"null" throws an exception).

I think there are 3 options

1. `"null"` → `JsonNull`
2. `"null"` → `throw JsonEncodingException`
3. `"null"` → `JsonLiteral("null", isString = false)`

You say Option 1 can cause problems, Option 2 will prevent problems, but I think Option 3 is fine, and should work as expected.

Would 3 work?


```kotlin
@ExperimentalSerializationApi
@Suppress("FunctionName")
public fun JsonUnquotedLiteral(value: String?): JsonPrimitive {
    return when (value) {
        null -> JsonNull
        else -> JsonLiteral(value, isString = false, coerceToInlineType = jsonUnquotedLiteralDescriptor)
    }
}
```

qwwdfsad replied:
Here is the test that expected to "fail" (if one replaces `println` with actual asserts):

```
@Test
fun testThatNullAndNullStringAreDistinguishable() {
    val withNull = buildJsonObject { put("key", JsonNull) }
    val withNullString = buildJsonObject { put("key", JsonUnquotedLiteral("null")) }
    println(withNull == withNullString) // 'false' as expected

    val withNullSerilized = Json.encodeToString(withNull)
    val withNullStringSerialized = Json.encodeToString(withNullString)

    println(withNullSerilized == withNullStringSerialized) // Should be 'false', but is 'true

    val deserialized1 = Json.decodeFromString<JsonElement>(withNullSerilized)
    val deserialized2 = Json.decodeFromString<JsonElement>(withNullStringSerialized)
    println(deserialized1 == deserialized2) // Should be 'false', but is 'true'
}
```

We have two **semantically different** json objects, yet they are serialized to the same JSON string and, when parsed, are coerced into one json object, effectively losing information.

While looking harmless, debugging such an error in an enterprise-level software, especially where `JsonElement` is used indirectly when read (e.g. by polymorphic serializer or some man-in-the-middle that invokes `JsonDecoder.decodeJsonElement`)

aSemy replied:
Thanks for the test! I've added a bit more logging to go over why I'm confused...

(I disabled the `value == "null" -> throw Exception` for these tests)

### JsonObject.toString()

```kotlin
val withNull = buildJsonObject { put("key", JsonNull) }
val withNullString = buildJsonObject { put("key", JsonUnquotedLiteral("null")) }
println(
    """
    withNull:                   $withNull
    withNullString:             $withNullString
    withNull == withNullString: ${withNull == withNullString}
    """.trimIndent()
)
```

```text
withNull:                   {"key":null}
withNullString:             {"key":null}
withNull == withNullString: false
```

Which seems wrong to me... They look the same. I think it makes sense that `JsonUnquotedLiteral("null") == JsonNull`.


### Json.encodeToString 

```kotlin
val withNullEncoded = Json.encodeToString(JsonElement.serializer(), withNull)
val withNullStringEncoded = Json.encodeToString(JsonElement.serializer(), withNullString)
println(
    """
    withNullEncoded:                          $withNullEncoded
    withNullStringEncoded:                    $withNullStringEncoded
    withNullEncoded == withNullStringEncoded: ${withNullEncoded == withNullStringEncoded}
    """.trimIndent()
)
```

```text
withNullEncoded:                          {"key":null}
withNullStringEncoded:                    {"key":null}
withNullEncoded == withNullStringEncoded: true
```

This makes sense to me, they look the same. I don't see why it should be false. I'd expect that `JsonUnquotedLiteral("null")` produces outputs `null`.

### Json.decodeToString 

```kotlin
val withNullDecoded = Json.decodeFromString(JsonElement.serializer(), withNullEncoded)
val withNullStringDecoded = Json.decodeFromString(JsonElement.serializer(), withNullStringEncoded)

println(
    """
    withNullDecoded:                          $withNullDecoded
    withNullStringDecoded:                    $withNullStringDecoded
    withNullDecoded == withNullStringDecoded: ${withNullDecoded == withNullStringDecoded}
    """.trimIndent()
)
```

```text
withNullDecoded:                          {"key":null}
withNullStringDecoded:                    {"key":null}
withNullDecoded == withNullStringDecoded: true
```

Again, this also makes sense. They're both the same JSON content.

---

I think it's just difficult for me to imagine how this situation could happen. `JsonUnquotedLiteral()` is akin to disabling type-checking using `dynamic` in Kotlin/JS, so putting a minor safety net in for `"null"` seems unusual. The horse has already left the stable, as they say. 

I imagine that debugging issues basically be asking "are you using `JsonUnquotedLiteral()`? You probably shouldn't!" But you're the maintainers, not me! So if an exception on `"null"` makes it easier for you, then it makes sense.

Anyway, to wrap it up, I'll happily admit that I don't need to understand this! And I really want this PR ready as soon as possible, so I'll try and adapt the test you posted, and update the KDoc on `JsonUnquotedLiteral()`.

aSemy commented on formats/json-tests/commonTest/src/kotlinx/serialization/json/serializers/JsonRawElementTest.kt:
```diff
@@ -0,0 +1,187 @@
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class JsonRawElementTest : JsonTestBase() {
```
I've expanded the tests here, but I got pretty stuck. Some of the failure messages don't make sense to me.

```
Failed with streaming = STREAMING expected: kotlinx.serialization.json.JsonLiteral<"lorem"> but was: kotlinx.serialization.json.JsonLiteral<"lorem">
```

```
expected:<{ ="invalid key"}> but was:<{" "="invalid key"}>
```

```
Failed with streaming = STREAMING expected: kotlinx.serialization.json.JsonLiteral<[1,2,3]> but was: kotlinx.serialization.json.JsonArray<[1,2,3]>
```

Can someone help out? :)

sandwwraith replied:
Actually, I think these tests don't have to be checked for decoding, because in most cases they would be different.

For example, if you encode `JsonRawElement("[1,2,3]")`, naturally, it appears in JSON as an array and is parsed back as an array, thus the difference. I don't think we should do anything with that, so it's OK here to test encoding.

aSemy replied:
Thanks, done!

sandwwraith commented:
Also, I've just noted that the PR is opened against `master` branch, while all new PRs with code should be opened against `dev`. Can you change the branch please?

aSemy commented:
I've updated the PR to point at `dev`, and merged `dev` into this branch, but it looks like `dev` is missing commits from `master`. Can `master` be merged into `dev`?

aSemy commented:
Okay, I think I've addressed all the comments, so hopefully there's just minor points to fix! 

I've added some Knit documentation, but I ran of energy towards the end so please give it a proof-read.

sandwwraith commented on formats/json-tests/commonTest/src/kotlinx/serialization/json/serializers/JsonUnquotedLiteralTest.kt:
```diff
@@ -0,0 +1,137 @@
+    private fun assertUnquotedLiteralEncoded(inputValue: String) = parametrizedTest { mode ->
+        val unquotedElement = JsonUnquotedLiteral(inputValue)
+
+        assertEquals(
```
Given that this assert is not dependent on the mode, it will be executed multiple times without any changes (currently there are 4 modes, IIRC). I suggest just moving it outside of `parametrizedTest` lambda

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -76,13 +80,52 @@ public fun JsonPrimitive(value: String?): JsonPrimitive {
+/**
+ * Create a [JsonPrimitive] from the given string, without surrounding it in quotes.
+ *
+ * **Note:** Creating a literal unquoted value of `null` (as in, `value == "null"`) is forbidden. If you want to create
```
I'd move this note (and without the word 'Note') down, under 'Be aware that'... because it would likely be a rare event

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -76,13 +80,52 @@ public fun JsonPrimitive(value: String?): JsonPrimitive {
+ * * large numbers,
+ * * or complex JSON objects.
+ *
+ * Be aware that it is possible to create invalid JSON using this function!
```
It's better to keep documentation tone neutral, without exclamation marks

```suggestion
 * Be aware that it is possible to create invalid JSON using this function.
```

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -76,13 +80,52 @@ public fun JsonPrimitive(value: String?): JsonPrimitive {
 public fun JsonPrimitive(value: Nothing?): JsonNull = JsonNull
 
+/**
+ * Create a [JsonPrimitive] from the given string, without surrounding it in quotes.
```
```suggestion
 * Creates a [JsonPrimitive] from the given string, without surrounding it in quotes.
```

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -76,13 +80,52 @@ public fun JsonPrimitive(value: String?): JsonPrimitive {
+ * Be aware that it is possible to create invalid JSON using this function!
+ *
+ * @see JsonPrimitive This is the preferred method for encoding JSON primitives.
+ * @throws JsonEncodingException If `value == "null"`
```
```suggestion
 * @throws JsonEncodingException if `value == "null"`
```

because pop-ups in IDEA render these as the single sentence

sandwwraith commented on formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt:
```diff
@@ -76,13 +80,52 @@ public fun JsonPrimitive(value: String?): JsonPrimitive {
+ *
+ * Be aware that it is possible to create invalid JSON using this function!
+ *
+ * @see JsonPrimitive This is the preferred method for encoding JSON primitives.
```
```suggestion
 * @see JsonPrimitive is the preferred method for encoding JSON primitives.
```

sandwwraith reviewed (commented):
I like your write-up! Only small comments left to fix.

Also, the build indeed gives knit errors, so rerun it `./gradlew knit` and check afterward (`./gradlew knitCheck`). Ping me if there are any problems.

aSemy commented:
I saw that knit was on 0.3.0, I bumped it to the latest: 0.4.0.

sandwwraith commented on docs/json.md:
```diff
@@ -25,6 +25,9 @@ In this chapter, we'll walk through features of [JSON](https://www.json.org/json
   * [Decoding Json elements](#decoding-json-elements)
+  * [Encoding raw, literal Json content (experimental)](#encoding-raw-literal-json-content-experimental)
+    * [Serializing large decimal numbers](#serializing-large-decimal-numbers)
+    * [Using `JsonUnquotedLiteral` to encode `"null"` is forbidden](#using-jsonunquotedliteral-to-encode-"null"-is-forbidden)
```
It seems the issue is not fixes in 0.4.0 :( Perhaps you can rephrase it to avoid `"null"` mention

aSemy replied:
Yeah, I'll try and rephrase it. I've just made an issue & PR for Knit https://github.com/Kotlin/kotlinx-knit/issues/45

aSemy replied:
It's fixed now 👍

sandwwraith reviewed (approved):
Great job, thanks for implementing this important feature!

sandwwraith commented:
No, we don't usually publish snapshots. The next release is going to be around Kotlin 1.8.0 release, in a month or two
